openlibrary/catalog/marc/marc_binary.py

```diff
@@ -92,13 +113,28 @@ def __init__(self, data):
         if len(data) != length:
             raise BadLength("Record length %s does not match reported length %s." % (len(data), length))
         self.data = data
+        self.directory_end = data.find('\x1e')
+        if self.directory_end == -1:
+            raise BadMARC("MARC directory not found")
+
+    def iter_directory(self):
+        data = self.data
+        directory = data[24:self.directory_end]
+        if len(directory) % 12 != 0:
+            # directory is the wrong size
+            # sometimes the leader includes some utf-8 by mistake
+            directory = data[:self.directory_end].decode('utf-8')[24:]
+            if len(directory) % 12 != 0:
+                raise BadMARC("MARC directory invalid length")
+        iter_dir = (directory[i*12:(i+1)*12] for i in range(len(directory) / 12))
+        return iter_dir
```

Comment:
This looks like it returns a tuple, not an iterator as one would expect from the name.

Reply:
@tfmorris 
Good to check. It does look like a tuple, but it is a generator. 

```
>>> iter_dir = ('123456789'[i] for i in range(9))
>>> iter_dir
<generator object <genexpr> at 0x7f0bd3c312d0>
>>> iter_dir.next()
'1'
>>> iter_dir.next()
'2'
>>> iter_dir.next()
'3'
```
Can you think of a way to make it clearer? I had assumed it was an iterator because I had seen it used by the code as if it were one, but when I looked at that line I thought I must have been wrong.

Interestingly, it's the difference between parenthesis and square brackets that changes the type:
```
>>> type(['123456789'[i] for i in range(9)])
<type 'list'>
>>> type(('123456789'[i] for i in range(9)))
<type 'generator'>
```

Reply:
Aaahh, I see. It's not a tuple, but a parenthesized generator expression. My bad! I think it's OK as is.